### DISCUSSION TITLE
Can't connect Python program in UBUNTU with Oracle DB via oracledb library

### DISCUSSION THREAD

**Discussion by GianlucaLostuzzo** [oracle/python-oracledb#190] [Q&A]
Hi,
I'm facing an issue with the connection to the Oracle DB of my enterprise's ERP with a python program that uses the oracledb library.
I tried to connect to the DB via thin mode but I get this error:
raise exc_type(_Error(message)) from cause
oracledb.exceptions.NotSupportedError: DPY-3015: password verifier type 0x939 is not supported by python-oracledb in thin mode

I tried then to connect using the thick mode but I get this error instead:
oracledb.exceptions.DatabaseError: DPI-1047: Cannot locate a 64-bit Oracle Client library: "libnnz21.so: cannot open shared object file: No such file or directory".

I don't know what to do now, reading the docs I can't find the answer to my problem.

Thanks in advance.

**Comment by anthony-tuininga** [comment]
The first error is due to the use of the 10g password verifier. If the parameter `sec_case_sensitive_logon` has been set to the value `false` and can be returned to its default value of `true` then thin mode will be able to connect successfully. Another option is that the password hasn't been changed since the database was upgraded from 10g. Simply changing the password (even to the same value it has now) would be sufficient.

Regarding thick mode, you do have to install the Oracle Client libraries. Take a look at the [installation documentation](https://python-oracledb.readthedocs.io/en/latest/user_guide/installation.html) and especially note the discussion surrounding the Oracle Client libraries and the enabling of thick mode.

**Reply by GianlucaLostuzzo** [reply]
Thanks for the answer, I don't think it is a problem related to the 10g password verifier, because I can correctly connect to another Oracle DB hosted in my server.
I'll check the documentation about the thick mode.

**Reply by anthony-tuininga** [reply]
You're welcome. That error is telling you that the server is only *supplying* a 10g password verifier -- either due to configuration or because that is the only password verifier available for that particular user! You can take a look at the contents of the column `password_versions` in the view `dba_users` for the user in question. If you only see the 10g value you know that you probably only have to change the password so that the 11g and 12c verifiers can be created. If you see the 11g and 12c verifiers there already then that database has the `sec_case_sensitive_logon` parameter set. This configuration can be specific to a particular database so the fact that you can connect to another one without difficulty doesn't say anything about this particular database!

**Comment by cjbj** [comment]
The second error (`libnnz21.so: cannot open shared object file`) would indicate to me that you called `init_oracle_client()` with a `libDir` on Linux.  This won't work.  See https://python-oracledb.readthedocs.io/en/latest/user_guide/initialization.html#setting-the-oracle-client-library-directory-on-linux-and-related-platforms
